feat(chisel): improve inspector

DaniPopes opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation

Currently many expressions are missing from the `Type` enum which handles expressions to be inspected
Fixes #4130

TLDR:
- Can inspect all Solidity expressions, excluding tuples, with the correct type. This includes, but not limited to:
  - All [global variables](https://docs.soliditylang.org/en/latest/cheatsheet.html#global-variables), with correct abi.decode, type(T).max etc types
  - All type associated methods (array.length, function.address)
  - All assignments, which also get added to the final code
- Inspected arithmetic operations will return int if at least 1 of the operands is also an int
<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution

- Add missing expressions
- Properly handle functions, accesses and tuples
- Refactor global variable logic to another method
- Clean up 73 levels of indentation functions

Listed in the code


- ~~Continue inspecting if `inspectoor` variable fails~~
  - ~~Any tuple (or empty return) fails compilation due to it not being able to be encoded in `inspectoor`~~
  - This is getting already too big, I think this is better for another PR
- [x] ~~Encode array length and relative offset for dynamic arrays~~ Array literals as `Type::FixedArray`
- [x] If assignment expressions return Some and get "inspected", the assignment does not happen
- [x] Exhaustively match on `from_expression` (blocked by the above)
- [x] Use `Int(n)` for arithmetic operations

Might not be needed or feasible:
- (?) Rational numbers (no corresponding ParamType ?)

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

mattsse commented on chisel/src/executor.rs:
```diff
@@ -376,82 +395,129 @@ impl Type {
+                    let num = num.as_deref().and_then(parse_number_literal);
+                    if let Some(num) = num {
+                        // overflow check
+                        const MAX_U64_AS_U256: U256 = U256([0, 0, 0, usize::MAX as u64]);
+                        if num > MAX_U64_AS_U256 {
```
last time we changed U256 constants in ethers we got burned, so I'd appreciate a simple sanity check for this constant.

mattsse commented on cli/tests/it/cmd.rs:
```diff
@@ -368,7 +368,7 @@ forgetest_init!(can_clean_config, |prj: TestProject, mut cmd: TestCommand| {
     cmd.assert_non_empty_stdout();
```
please submit the clippy/style changes for `cli/tests` separately

mattsse reviewed (changes requested):
code-wise this lgtm, 
not too familiar with the issue, can we get a test case for this perhaps?

wdyt @clabby @rkrasiuk

DaniPopes commented:
Not sure how I am supposed to configure Solc to run tests ..

mattsse commented on cli/tests/it/cmd.rs:
```diff
@@ -368,7 +368,7 @@ forgetest_init!(can_clean_config, |prj: TestProject, mut cmd: TestCommand| {
     cmd.assert_non_empty_stdout();
 
```
didn't we just merge a PR that changed the format! s?

In any case, please undo because unrelated here.

DaniPopes replied:
this was reverted like 10 commits ago , this isnt the current diff

mattsse reviewed (commented):
a lot of unrelated changes which makes this large PR harder to review, please undo unrelated changes and submit separately if applicable.

clabby commented:
> code-wise this lgtm, not too familiar with the issue, can we get a test case for this perhaps?
> 
> wdyt @clabby @rkrasiuk

Busy with some work stuff, but will review this & #4147 later on tonight or tomorrow morning! Appreciate the improvements @DaniPopes 🚀

rkrasiuk commented:
same as what @mattsse said, the code looks good. deferring to @clabby for approval

x676f64 commented on chisel/src/executor.rs:
```diff
@@ -458,15 +458,15 @@ impl Type {
             // pt::Expression::PostDecrement(_, inner) |       // <inner>--
             // pt::Expression::PreIncrement(_, inner) |        // ++<inner>
             // pt::Expression::PostIncrement(_, inner) |       // <inner>++
-            // pt::Expression::Assign(_, inner, _) |           // <inner> = ...
```
Should remove this change from your commit.

x676f64 commented:
> code-wise this lgtm, not too familiar with the issue, can we get a test case for this perhaps?
> 
> wdyt @clabby @rkrasiuk

The initial impetus for this PR was https://github.com/foundry-rs/foundry/issues/4130 which includes some failing test cases pertaining to signed integers.

clabby reviewed (commented):
This is looking pretty good code-wise, but experiencing some issues with inspecting on commit `76acc498ea29cbddf5b649467ff25131b0ffcf2b`. On my machine, it crashes every time a variable or expression is inspected (regardless of the expression type)

Env:
- **OS:** Mac OS
- **Arch**: ARM (Apple Silicon)
<img width="554" alt="Screenshot 2023-01-24 at 2 40 14 PM" src="https://user-images.githubusercontent.com/8406232/214309692-d9bef755-1fde-48b0-8eff-d92134445009.png">

mattsse commented:
we merged two other chisel PRs yesterday,

just to rule it out, this does __not__ panic with nightly chisel?

DaniPopes commented:
> On my machine, it crashes every time a variable or expression is inspected (regardless of the expression type)

I introduced infinite recursion in 799e2b42653fa88b27d9bb92e404b430dfc3611f by mistake because I didn't merge the `try_as_ethabi` match arms properly. Fixed in ed8683863ba6d5a24659241bd8cc0699bdf48367, sorry about that

clabby reviewed (approved):
This looks good to me, thanks for the improvements! There are some commented out blocks (specifically in the `executor` mod) that could use some cleaning up, but that's just a stylistic nit.

mattsse commented on chisel/src/executor.rs:
```diff
@@ -657,100 +888,288 @@ impl Type {
+        expr: &pt::Expression,
+        intermediate: Option<&IntermediateOutput>,
+        custom_type: &mut Vec<String>,
+        // contract_name: Option<String>,
+        // intermediate_contract: Option<&IntermediateContract>,
```
no longer used? can remove?

mattsse commented on chisel/src/executor.rs:
```diff
@@ -657,100 +888,288 @@ impl Type {
+                let name = &ident.name;
+
+                if let Some(intermediate) = intermediate {
+                    // if let Some(intermediate_contract) = intermediate_contract {
+                    //     if intermediate_contract.struct_definitions.contains_key(name) {
+                    //         custom_type.clear();
+                    //         custom_type.push(name.clone());
+                    //         return Self::infer_custom_type(intermediate, custom_type,
+                    // contract_name)     }
+                    // }
```
can be removed?

mattsse reviewed (approved):
pending, code comment nits otherwise lgtm

tysm

mattsse reviewed (approved):
gg champ
